rust/kernel/types.rs

```diff
@@ -295,6 +297,111 @@ opaque_init_funcs! {
     "Rust" manual_init4(arg1: A1, arg2: A2, arg3: A3, arg4: A4);
 }
 
+/// Types that are _always_ reference counted.
+///
+/// It allows such types to define their own custom ref increment and decrement functions.
+/// Additionally, it allows users to convert from a shared reference `&T` to an owned reference
+/// [`ARef<T>`].
+///
+/// This is usually implemented by wrappers to existing structures on the C side of the code. For
+/// Rust code, the recommendation is to use [`Arc`](crate::sync::Arc) to create reference-counted
+/// instances of a type.
+///
+/// # Safety
+///
+/// Implementers must ensure that increments to the reference count keep the object alive in memory
+/// at least until matching decrements are performed.
+///
+/// Implementers must also ensure that all instances are reference-counted. (Otherwise they
+/// won't be able to honour the requirement that [`AlwaysRefCounted::inc_ref`] keep the object
+/// alive.)
+pub unsafe trait AlwaysRefCounted {
```

Comment:
Not sure I get the name. Do we have objects that are only some times refcounted?

How about `ForeignRefCounted` in line with `ForeignOwnable`?

Reply:
Not sure about that name. It doesn't have to be foreign.

`Always` here means that all instances of `Self` must be refcounted -- thus you can turn `&Self` into `ARef<Self>`. It is intended to contrast with `Arc`, which doesn't allow you to turn `&T` into `Arc<T>`.

We have bikeshedded the name in the last meeting, between `AlwaysRefCounted` and `RefCounted`, and `ARef` vs `Orc/Irc/...`, and we've decided to keep using the current name via polling.